### PR TITLE
Enable stylecheck linter

### PR DESCRIPTION
**Pull request by ankitjain235**
## Change Overview

This PR enables stylecheck golang linter

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -75,4 +76,6 @@ linters-settings:
     sections:
       - standard
       - default
-      - prefix(github.com/kanisterio/kanister)
```
Comment (PrasadG193) [inline .golangci.yml:78]: Extra newlines?

File: pkg/controller/metrics.go [hunk 2]
```diff
@@ -27,26 +27,26 @@ type metrics struct {
 }
 
 const (
-	ACTION_SET_COUNTER_VEC_LABEL_RES         = "resolution"
-	ACTION_SET_COUNTER_VEC_LABEL_RES_SUCCESS = "success"
-	ACTION_SET_COUNTER_VEC_LABEL_RES_FAILURE = "failure"
+	ACTION_SET_COUNTER_VEC_LABEL_RES         = "resolution" //nolint:stylecheck // Already defined in docs
+	ACTION_SET_COUNTER_VEC_LABEL_RES_SUCCESS = "success"    //nolint:stylecheck // Already defined in docs
+	ACTION_SET_COUNTER_VEC_LABEL_RES_FAILURE = "failure"    //nolint:stylecheck // Already defined in docs
```
Comment (PrasadG193) [inline pkg/controller/metrics.go:32]: What is the error we are getting?

File: pkg/kopia/command/maintenance.go [hunk 3]
```diff
@@ -16,7 +16,7 @@ package command
 
 type MaintenanceInfoCommandArgs struct {
 	*CommandArgs
-	GetJsonOutput bool
+	GetJsonOutput bool //nolint:stylecheck // Validate workflows before enabling stylecheck
```
Comment (PrasadG193) [inline pkg/kopia/command/maintenance.go:19]: What is the error we are getting here? Why do we need to ignore this?

File: pkg/kopia/command/storage/s3_args.go [hunk 4]
```diff
@@ -38,7 +38,7 @@ func s3Args(location map[string][]byte, repoPathPrefix string) logsafe.Cmd {
 		s3Endpoint := ResolveS3Endpoint(e)
 		args = args.AppendLoggableKV(s3EndpointFlag, s3Endpoint)
 
-		if HttpInsecureEndpoint(e) {
+		if HTTPInsecureEndpoint(e) {
```
Comment (PrasadG193) [inline pkg/kopia/command/storage/s3_args.go:41]: If its not used outside the pkg, can we make it internal/private?

File: pkg/kube/pod.go [hunk 5]
```diff
@@ -442,9 +442,9 @@ func patchDefaultPodSpecs(defaultPodSpecs corev1.PodSpec, override crv1alpha1.JS
 }
 
 // CreateAndMergeJsonPatch uses Strategic Merge to merge two Pod spec configuration
-func CreateAndMergeJsonPatch(original, override crv1alpha1.JSONMap) (crv1alpha1.JSONMap, error) {
+func CreateAndMergeJsonPatch(original, override crv1alpha1.JSONMap) (crv1alpha1.JSONMap, error) { //nolint:stylecheck // Validate workflows before enabling stylecheck
```
Comment (PrasadG193) [inline pkg/kube/pod.go:445]: Why do we need to ignore the check?

File: pkg/version/version.go [hunk 6]
```diff
@@ -19,9 +19,10 @@ package version
 import "fmt"
 
 // The below variables will be overrriden using ldflags set by goreleaser during the build process
+
 var VERSION = "DEV"
-var GIT_COMMIT = "NONE"
-var BUILD_DATE = "UNKNOWN"
+var GIT_COMMIT = "NONE"    //nolint:stylecheck // Validate goreleaser works find after the change
+var BUILD_DATE = "UNKNOWN" //nolint:stylecheck // Validate goreleaser works find after the change
```
Comment (PrasadG193) [inline pkg/version/version.go:25]: Why do we need to ignore the checks? We can add the reason in the comment?

Comment (PrasadG193) [reply pkg/version/version.go:25]: `// Validate goreleaser works find after the change` Is it a TODO comment? Something we are supposed to do as a part of this PR? I am assuming it will be removed once we validate that, right>

File: pkg/kopia/command/policy_show_global.go [hunk 7]
```diff
@@ -16,7 +16,7 @@ package command
 
 type PolicyShowGlobalCommandArgs struct {
 	*CommandArgs
-	GetJsonOutput bool
+	GetJsonOutput bool //nolint:stylecheck // Validate workflows before enabling stylecheck
```
Comment (PrasadG193) [inline pkg/kopia/command/policy_show_global.go:19]: @ankitjain235 Is it a todo comment? The explanation is not clear

File: pkg/controller/metrics.go (hunk 2 above)
Comment (pavannd1) [reply pkg/controller/metrics.go:32]: Could you double-check if the metrics setting is enabled by default? Also, the label values are going to be the same right? We would only be changing the constant used to refer to them

**Comment by pavannd1** [comment]
Not stale

File: pkg/version/version.go (hunk 6 above)
Comment (saima-s) [reply pkg/version/version.go:25]: `GIT_COMMIT` and `BUILD_DATE` is referred only in VersionString functions [here](https://github.com/kanisterio/kanister/pull/2505/files#diff-73b8ec8a05213857da6e9598e34c5d7a18a6e5349250908ae6710578236977f2L28) and its value being used. Do we really need to test if goreleaser works fine? Beacuse value as `NONE` and `UNKNOWN` will be used in reference to the var names.

Comment (ankitjain235) [reply pkg/version/version.go:25]: These fields do seen to be used from goreleaser https://github.com/kanisterio/kanister/blob/1ebea1abd67ab4c8070a4385d9609f125438c33e/.goreleaser.yml#L11-L13

Comment (saima-s) [reply pkg/version/version.go:25]: ACK.Checking the testing method.

File: pkg/kopia/command/policy_show_global.go (hunk 7 above)
Comment (saima-s) [reply pkg/kopia/command/policy_show_global.go]: refactored

File: pkg/controller/metrics.go (hunk 2 above)
Comment (saima-s) [reply pkg/controller/metrics.go]: checked. Only `ACTION_SET_COUNTER_VEC_LABEL_RES` variable is used in documentation, others are not referenced anywhere but code. Refactored those.

File: pkg/validate/validate.go [hunk 8]
```diff
@@ -72,7 +72,7 @@ func actionSpec(s crv1alpha1.ActionSpec) error {
 		// Not a known type. ActionSet must specify API group and resource
 		// name in order to populate `Object` TemplateParam
 		if s.Object.APIVersion == "" || s.Object.Resource == "" {
-			return errorf(validateErr, "Not a known object Kind %s. Action %s must specify Resource name and API version", s.Object.Kind, s.Name)
+			return errorf(errValidate, "Not a known object Kind %s. Action %s must specify Resource name and API version", s.Object.Kind, s.Name)
```
Comment (viveksinghggits) [inline pkg/validate/validate.go:75]: does `validateErr` not pass but `errValidate` pass?

File: pkg/controller/metrics.go [hunk 9]
```diff
@@ -27,46 +27,46 @@ type metrics struct {
 }
 
 const (
-	ACTION_SET_COUNTER_VEC_LABEL_RES         = "resolution"
-	ACTION_SET_COUNTER_VEC_LABEL_RES_SUCCESS = "success"
-	ACTION_SET_COUNTER_VEC_LABEL_RES_FAILURE = "failure"
+	ACTION_SET_COUNTER_VEC_LABEL_RES   = "resolution" //nolint:stylecheck // Already defined in docs
```
Comment (pavannd1) [inline pkg/controller/metrics.go:30]: 1. Let's verify if the metrics are enabled by default controller install or through a separate helm flag.
2. If it's through a helm flag, then it would be okay to change the documentation.
3. If it's enabled by default, what is the impact of the change? We should have a plan for the transition - a release note about the breaking change and actual change in a subsequent release.
@viveksinghggits @hairyhum wdyt?

Comment (hairyhum) [reply pkg/controller/metrics.go:30]: Metrics are enables via helm flag `metrics.enabled`, but we don't configure labels via helm, as I understand.
I don't see why we shouldn't change it. The variable name is not an input anywhere and I can only find it in `kanister_prometheus_integration.md` which describes how it works but does not suggest you use it.
Is there some other docs which mention this variable?

File: pkg/validate/validate.go (hunk 8 above)
Comment (saima-s) [reply pkg/validate/validate.go:75]: yes, aacording to stylecheck lint(ST1012), it says `error var validateErr should have name of the form errFoo (stylecheck)`

File: pkg/controller/metrics.go (hunk 9 above)
Comment (saima-s) [reply pkg/controller/metrics.go:30]: In `kanister_prometheus_integration.md` there are other 2 variables `ACTION_SET_COUNTER_VEC_LABEL_OP_TYPE` and `ACTION_SET_COUNTER_VEC_LABEL_BLUEPRINT` , there is no occurrence of these 2 in codebase apart from this doc. I am not sure what is the use of those and whether to we should rename it the same way as we are doing it for `ACTION_SET_COUNTER_VEC_LABEL_RES` to `ActionSetCounterVecLabelRes`. It seems inconsistent in doc if we don't do.

Comment (hairyhum) [reply pkg/controller/metrics.go]: It seems like `kanister-prometheus-integration.md` is outdated as we don't have `ACTION_SET_COUNTER_VEC_LABEL_OP_TYPE` in the code anymore. 
Maybe we should make it consistent with the code or don't have code in `md` files. We can do that in a separate PR of course.

File: pkg/app/couchbase.go [hunk 10]
```diff
@@ -52,6 +52,8 @@ type CouchbaseDB struct {
 }
 
 // Last tested woking version "2.3.0"
+//
+//nolint:stylecheck
```
Comment (viveksinghggits) [inline pkg/app/couchbase.go:56]: why did we have to add the `nolint` here?

**Review by viveksinghggits — CHANGES_REQUESTED** [review]

File: pkg/consts/consts.go [hunk 11]
```diff
@@ -47,6 +48,7 @@ const (
 )
 
 // These consts are used to query Repository server API objects
+
```
Comment (viveksinghggits) [inline pkg/consts/consts.go:51]: is this required? i would prefer is removed.

File: pkg/metrics/metrics.go [hunk 12]
```diff
@@ -11,6 +11,7 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+
```
Comment (viveksinghggits) [inline pkg/metrics/metrics.go:14]: I see that in some source files we are adding a new line immediately before the `package` declaration and in some files not (example `aws_secrets.go`) why is that?

File: pkg/testing/integration_register.go [hunk 13]
```diff
@@ -85,7 +85,7 @@ var _ = Suite(&CockroachDB{
 	},
 })
 
-// time-log app for csi volumesnapshot
+// TimeLogCSI registers time-log app for csi volumesnapshot.
```
Comment (viveksinghggits) [inline pkg/testing/integration_register.go:88]: I think these comments should be improved, these are just types they don't register the app right?

File: pkg/testing/integration_register.go [hunk 14]
```diff
@@ -279,7 +279,7 @@ var _ = Suite(&MysqlDBDepConfig{
 	},
 })
 
-// MongoDB deployed on openshift cluster
+// MongoDBDepConfig deploys MongoDB on openshift cluster
```
Comment (viveksinghggits) [inline pkg/testing/integration_register.go:282]: It's just a type, it doesn't deploy the app.

**Review by viveksinghggits — CHANGES_REQUESTED** [review]

File: pkg/app/couchbase.go (hunk 10 above)
Comment (saima-s) [reply pkg/app/couchbase.go:56]: This is because style check expects exported function should have description of what its doing.
```
ST1020: comment on exported function NewCouchbaseDB should be of the form "NewCouchbaseDB ...
```

File: pkg/metrics/metrics.go (hunk 12 above)
Comment (saima-s) [reply pkg/metrics/metrics.go:14]: There is always a line after `// limitations under the License.`
incase of `aws_secrets.go`, there a package description ` Package repositoryserver validates the secret credentials for all location profiles.` written before `package` (for style check linter ST1000) which is a different issue to be solved in another PR(as I am not aware of all the package descriptions)

File: pkg/app/couchbase.go (hunk 10 above)
Comment (viveksinghggits) [reply pkg/app/couchbase.go:56]: I think instead of `nolint` we should just add `NewCouchbaseDB initialises an instance of Couchbase DB` or something in these line. We are anyway adding the comment of `last tested` and a new line after `last tested` doesn't make sense.

File: pkg/metrics/metrics.go (hunk 12 above)
Comment (viveksinghggits) [reply pkg/metrics/metrics.go:14]: ok

File: pkg/testing/integration_register.go (hunk 13 above)
Comment (saima-s) [reply pkg/testing/integration_register.go]: done

File: pkg/testing/integration_register.go (hunk 14 above)
Comment (saima-s) [reply pkg/testing/integration_register.go]: done

File: pkg/consts/consts.go (hunk 11 above)
Comment (saima-s) [reply pkg/consts/consts.go:55]: refactored.

File: pkg/app/couchbase.go (hunk 10 above)
Comment (saima-s) [reply pkg/app/couchbase.go]: done

File: pkg/testing/integration_register.go [hunk 15]
```diff
@@ -115,7 +115,7 @@ var _ = Suite(&Maria{
 	},
 })
 
-// Elasticsearch app
+// Elasticsearch type is an app for elasticsearch for integration test..
```
Comment (viveksinghggits) [inline pkg/testing/integration_register.go:118]:
```suggestion
// Elasticsearch type is an app for elasticsearch for integration test.
```

**Review by viveksinghggits — APPROVED** [review]

File: design/kanister-prometheus-integration.md [hunk 16]
```diff
@@ -215,7 +215,7 @@ controller/metrics.go
 
 ```golang
 const (
-	ACTION_SET_COUNTER_VEC_LABEL_RES     = "resolution"
```
Comment (PrasadG193) [inline design/kanister-prometheus-integration.md:218]: [nit] I think we should exclude `./design` directory from the link checks

File: pkg/secrets/repositoryserver/aws_secrets.go [hunk 17]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// Package repositoryserver validates the secret credentials for all location profiles.
```
Comment (PrasadG193) [inline pkg/secrets/repositoryserver/aws_secrets.go:15]: [question] Do we have a lint checker for comment on package name? I don't think that is necessary

File: pkg/version/version.go [hunk 18]
```diff
@@ -19,12 +19,13 @@ package version
 import "fmt"
 
 // The below variables will be overrriden using ldflags set by goreleaser during the build process
-var VERSION = "DEV"
-var GIT_COMMIT = "NONE"
-var BUILD_DATE = "UNKNOWN"
+
+var Version = "DEV"
+var GitCommit = "NONE"
+var BuildDate = "UNKNOWN"
```
Comment (PrasadG193) [inline pkg/version/version.go:25]: This would break the release. We have to update ldflags in `.goreleaser` as per the changes. https://github.com/kanisterio/kanister/blob/master/.goreleaser.yml#L15

**Review by PrasadG193 — CHANGES_REQUESTED** [review]

File: pkg/version/version.go (hunk 18 above)
Comment (PrasadG193) [reply pkg/version/version.go:25]: Oh, right. My bad

**Review by PrasadG193 — APPROVED** [review]

File: pkg/secrets/repositoryserver/aws_secrets.go (hunk 17 above)
Comment (saima-s) [reply pkg/secrets/repositoryserver/aws_secrets.go:15]: Yes, it is. ST1000(stylecheck linter). It says we should describe package about what its doing in any one file atleast. Although I have disabled this `ST1000` in this PR, but had already made this change, so decided to keep the description as it is and later on `ST1000` will be enabled as a part of another PR will add similar description for other packages as well.

**Comment by viveksinghggits** [comment]
@saima-s let's merge 🚀 it.

**Comment by saima-s** [comment]
> @saima-s let's merge 🚀 it.

Will merge with https://github.com/kastenhq/k10/pull/26456/files.

File: pkg/secrets/repositoryserver/aws_secrets.go (hunk 17 above)
Comment (saima-s) [reply pkg/secrets/repositoryserver/aws_secrets.go:15]: @PrasadG193 Checked kubernates repository. Found few examples where this linter is handled:
1. https://github.com/kubernetes/kubernetes/blob/0fc167103189a4d462c3cc7a17b360c3d998f4bf/cmd/kube-apiserver/app/options/options.go#L17
2. https://github.com/kubernetes/kubernetes/blob/0fc167103189a4d462c3cc7a17b360c3d998f4bf/cmd/kube-apiserver/app/aggregator.go#L17
3. https://github.com/kubernetes/kubernetes/blob/0fc167103189a4d462c3cc7a17b360c3d998f4bf/cmd/kube-apiserver/apiserver.go#L17

File: design/kanister-prometheus-integration.md (hunk 16 above)
Comment (saima-s) [reply design/kanister-prometheus-integration.md:218]: This is not checking linter on `./design` package. The change ActionSetCounterVecLabelRes is made because this is the variable referred in code, and there changes needs to be done.
